Fix panic in AWS ECS detector if container ARN is not valid

DarthSim opened the pull request:
This PR fixes the panic that may occur here: https://github.com/open-telemetry/opentelemetry-go-contrib/blob/main/detectors/aws/ecs/ecs.go#L116

If `containerMetadata.ContainerARN` doesn't include `:` (that is our client case), `strings.LastIndex(containerMetadata.ContainerARN, ":")` will return `-1`. This leads to `containerMetadata.ContainerARN[:-1]`, and `index out of range` panic.

Here's what changes in this PR do:
- Try to extract `baseArn` from `taskMetadata.TaskARN', 'containerMetadata.ContainerARN', and 'taskMetadata.Cluster`
- If `baseArn` is successfully extracted, and some of the ARNs aren't full, fix these ARNs

pellared commented on CHANGELOG.md:
```diff
@@ -12,6 +12,9 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 
 - Add the new `go.opentelemetry.io/contrib/instrgen` package to provide auto-generated source code instrumentation. (#3068)
 
+### Fixed
+- Fix panic in `go.opentelemetry.io/contrib/detectors/aws/ecs` when the container ARN is not valid.
```
```suggestion
### Fixed

- Do not panic in `go.opentelemetry.io/contrib/detectors/aws/ecs` when the container ARN is not valid.
```

DarthSim commented on detectors/aws/ecs/ecs.go:
```diff
@@ -101,20 +101,28 @@ func (detector *resourceDetector) Detect(ctx context.Context) (*resource.Resourc
+			taskMetadata.Cluster,
+		)
+
+		if len(baseArn) > 0 {
```
We check string emptiness using `len` here: https://github.com/open-telemetry/opentelemetry-go-contrib/blob/main/detectors/aws/ecs/ecs.go#L81. Since there is no difference in speed between `len(str) > 0` and `str != ""`, I believe it's better to maintain the code style. WDYT?

pellared replied:
> I believe it's better to maintain the code style

I agree and `str != ""` is used in all (for sure most) other places (e.g. `gcp` detector). [This line](https://github.com/open-telemetry/opentelemetry-go-contrib/blob/aead73790726a6f6c8cc5facca0c7e43753399aa/detectors/aws/ecs/ecs.go#L81) should be changed as well.

DarthSim commented on detectors/aws/ecs/ecs.go:
```diff
@@ -128,7 +136,8 @@ func (detector *resourceDetector) Detect(ctx context.Context) (*resource.Resourc
 		attributes = append(
 			attributes,
-			semconv.AWSECSClusterARN(clusterArn),
+			semconv.AWSECSContainerARN(containerMetadata.ContainerARN),
```
I agree, but I just moved this string from above

pellared commented on detectors/aws/ecs/ecs.go:
```diff
@@ -101,20 +101,28 @@ func (detector *resourceDetector) Detect(ctx context.Context) (*resource.Resourc
+			if !strings.HasPrefix(taskMetadata.Cluster, "arn:") {
+				taskMetadata.Cluster = fmt.Sprintf("%s:cluster/%s", baseArn, taskMetadata.Cluster)
+			}
+			if !strings.HasPrefix(containerMetadata.ContainerARN, "arn:") {
+				containerMetadata.ContainerARN = fmt.Sprintf("%s:container/%s", baseArn, containerMetadata.ContainerARN)
+			}
+			if !strings.HasPrefix(taskMetadata.TaskARN, "arn:") {
+				taskMetadata.TaskARN = fmt.Sprintf("%s:task/%s", baseArn, taskMetadata.TaskARN)
+			}
```
> I'd say that it fixes the ARNs. 

I just say it is not mentioned in the changelog (it is only a matter of updating the changlog). It only says that there is no panic for improper container ARN. While the truth is that it does more 😉

> I can add the same test for TaskARN

That is what I suggest doing 👍

DarthSim commented:
Hey everyone!
Sorry for the long delay. I updated the PR:
- Added an entry about the task ARN to the changelog.
- Added test for fixing the task ARN.
- Applied suggested changes.

DarthSim commented:
Hey @pellared @Aneurysm9!

Is there any chance for this PR to be merged? Some of our users face the issue fixed by this PR, so I have to use a fork of this package and update it every time I update the project dependencies; this is pretty annoying.

Aneurysm9 commented on detectors/aws/ecs/test/ecs_test.go:
```diff
@@ -91,6 +91,120 @@ func TestDetectV4LaunchTypeEc2(t *testing.T) {
+	}))
+	defer testServer.Close()
+
+	os.Clearenv()
+	_ = os.Setenv(metadataV4EnvVar, testServer.URL)
```
Don't we have test helpers for setting environment variables without stepping on other tests?

Looks like [these](https://github.com/open-telemetry/opentelemetry-go/blob/main/internal/internaltest/env.go) may only be available in the core repo.  It would be good to be able to use them in this repo as well.  @pellared can you create an issue to track this if you agree?

pellared replied:
The tests could simply use https://pkg.go.dev/testing#T.Setenv
